### ISSUE TITLE
incorrect duplicate key error from sysbench read-write

### ISSUE THREAD

**Issue by mdcallag** [facebook/mysql-5.6#482]
tl;dr - I am getting a dup key error from sysbench with MyRocks and I don't think I should be getting it.

I have been running a few sysbench tests with all combinations of the following options (or option groups). There are 4 option groups and each is off or on, so there are 16 combinations:
* rocksdb_disable_2pc
* concurrent memtable -  rocksdb_allow_concurrent_memtable_write, rocksdb_enable_write_thread_adaptive_yield
* binlog_order_commits
* sync_binlog

For InnoDB I tested 4 configurations, the cross product of binlog_order_commits and sync_binlog.
I used a MyRocks build from November 21 and 8 tables with 1M rows/table for sysbench. The sysbench client was run for 5 minutes at these concurrency levels: 1 2 4 8 16 24 32 40 48 64 80 96 128. The test sequence is prepare, update-only, read-write, insert-only. For 2 of the 16 MyRocks configurations I get a dup key error on the insert during the read-write test. This doesn't occur for any of the 4 InnoDB configurations. I repeated the result (errors in the 2 configs) with sysbench running for 10 minutes at each level of concurrency and these concurrency levels: 1 8 32 64 128 256 512 1024. There is no error for 14 of the 16 MyRocks configs and there is no error for InnoDB so I doubt this is an InnoDB bug.

The 2 configurations for which I get the error are: 2pc enabled, concurrent memtable enabled, binlog_order_commits disabled and sync_binlog enabled/disabled. The failure only occurs at high concurrency -- 64 or more clients on server with 48 cores after HT was enabled.

The smallest reproduction test is:
```
nsecs=$(( 3600 * 24 )); nt=64; ./sysbench --test=tests/db/oltp.lua --db-driver=mysql --mysql-engine-trx=yes --mysql-user=root --mysql-password=pw --mysql-host=127.0.0.1 --mysql-db=test --mysql-table-engine=rocksdb --oltp-range-size=100 --oltp-table-size=1000000 --oltp-tables-count=8 --num-threads=$nt --max-requests=0 --max-time=$nsecs --oltp-read-only=off --oltp-index-updates=1 --oltp-non-index-updates=1 --oltp-delete-inserts=1 prepare

nsecs=$(( 3600 * 24 )); nt=64; ./sysbench --test=tests/db/oltp.lua --db-driver=mysql --mysql-engine-trx=yes --mysql-user=root --mysql-password=pw --mysql-host=127.0.0.1 --mysql-db=test --mysql-table-engine=rocksdb --oltp-range-size=100 --oltp-table-size=1000000 --oltp-tables-count=8 --num-threads=$nt --max-requests=0 --max-time=$nsecs --oltp-read-only=off --oltp-index-updates=1 --oltp-non-index-updates=1 --oltp-delete-inserts=1 run
```

**Comment by mdcallag** [comment]
This might be the same bug as https://github.com/facebook/mysql-5.6/issues/481
